public/app/plugins/datasource/loki/querybuilder/operationUtils.ts

```diff
@@ -157,6 +157,33 @@ export function labelFilterRenderer(model: QueryBuilderOperation, def: QueryBuil
   return `${innerExpr} | ${model.params[0]} ${model.params[1]} \`${model.params[2]}\``;
 }
 
+export function isConflictingFilter(
+  operation: QueryBuilderOperation,
+  queryOperations: QueryBuilderOperation[]
+): boolean {
+  let conflictingFilter = false;
+
+  const labelFilters = queryOperations.filter((op) => {
+    return op.id === LokiOperationId.LabelFilter && op.params !== operation.params;
+  });
+
+  // check if the new filter conflicts with any other label filter
+  labelFilters.forEach((filter) => {
+    if (operation.params[0] !== filter.params[0] || operation.params[2] !== filter.params[2]) {
+      return;
+    }
+
+    if (
+      (String(operation.params[1]).startsWith('!') && !String(filter.params[1]).startsWith('!')) ||
+      (String(filter.params[1]).startsWith('!') && !String(operation.params[1]).startsWith('!'))
+    ) {
+      conflictingFilter = true;
+    }
+  });
+
+  return conflictingFilter;
+}
```

Comment:
An alternative approach to consider.

```suggestion
export function isConflictingFilter(
  operation: QueryBuilderOperation,
  queryOperations: QueryBuilderOperation[]
): boolean {
  const operationIsNegative = operation.params[1].toString().startsWith('!');
  // Discard non label filters and different params
  const candidates = queryOperations.filter(queryOperation => (queryOperation.id === LokiOperationId.LabelFilter && queryOperation.params[0] === operation.params[0] && queryOperation.params[2] === operation.params[2]));
  
  const conflict = candidates.find((candidate) => {
    if (operationIsNegative && candidate.params[1].toString().startsWith('!') === false) {
      return true;
    }
    if (operationIsNegative === false && candidate.params[1].toString().startsWith('!')) {
      return true;
    }
    return false;
  });

  return conflict !== undefined;
}
```

A limitation of the current implementation is that it will continue iterating over all the label filters, even if one has already been found in line 180. I would suggest to change the `forEach` to a `find` or `findIndex`.

Reply:
Definitely. And it looks like it's potentially faster https://www.measurethat.net/Benchmarks/Show/4337/0/array-find-vs-some

Reply:
@matyax cool website, based off the fact `some` is faster, ill update to use that.